### PR TITLE
Fix bug of Checkpoint loses recent transactions with 2PC

### PR DESCRIPTION
**Pull request by siying**
Summary:
If 2PC is enabled, checkpoint may not copy previous log files that contain uncommitted prepare records. In this diff we keep those files.

Test Plan: Add a unit test to cover we copy just the log files we needed

### CODE REVIEW

**Comment by siying** [comment]
I'm using @yoshinorim 's test case. Modified a little bit to make sure we don't copy too many files.

File: utilities/checkpoint/checkpoint.cc [hunk 1]
```diff
@@ -62,6 +62,8 @@ Status CheckpointImpl::CreateCheckpoint(const std::string& checkpoint_dir) {
   Status s;
   std::vector<std::string> live_files;
   uint64_t manifest_file_size = 0;
+  bool allow_2pc = db_->GetDBOptions().allow_2pc;
+  uint64_t min_log_number_to_keep = port::kMaxUint64;
```
Comment (IslamAbdelRahman) [inline utilities/checkpoint/checkpoint.cc]: can we add a comment to explain why we need `min_log_number_to_keep `

File: utilities/checkpoint/checkpoint.cc [hunk 2]
```diff
@@ -78,6 +80,20 @@ Status CheckpointImpl::CreateCheckpoint(const std::string& checkpoint_dir) {
...
     s = db_->GetLiveFiles(live_files, &manifest_file_size);
+
+    if (s.ok() && allow_2pc) {
+      // If 2PC is enabled, we need to get minimum log number after the flush.
+      // Need to refetch the live files to recapture the snapshot.
+      if (!db_->GetIntProperty(DB::Properties::kMinLogNumberToKeep,
+                               &min_log_number_to_keep)) {
+        db_->EnableFileDeletions(false);
+        return Status::InvalidArgument(
+            "2PC enabled but cannot fine the min log number to keep.");
```
Comment (IslamAbdelRahman) [inline utilities/checkpoint/checkpoint.cc]: `find` instead of `fine`

File: utilities/checkpoint/checkpoint.cc [hunk 3]
```diff
@@ -78,6 +80,20 @@ Status CheckpointImpl::CreateCheckpoint(const std::string& checkpoint_dir) {
...
+      }
+      // No need to flush this time.
+      s = db_->GetLiveFiles(live_files, &manifest_file_size, false);
```
Comment (IslamAbdelRahman) [inline utilities/checkpoint/checkpoint.cc]: I am a bit confused, why do we need to call `GetLiveFiles` again

File: utilities/checkpoint/checkpoint.cc (hunk 1 above)
Comment (IslamAbdelRahman) [inline utilities/checkpoint/checkpoint.cc]: Can we also fix unity? It's complaining that this variable shadow's `min_log_number_to_keep` in `db/internal_stats.cc`

Comment (siying) [inline utilities/checkpoint/checkpoint.cc]: Sure.

**Comment by siying** [comment]
@IslamAbdelRahman  Fix unity, change the second call of GetLiveFiles() to force flush and add comments to explain why.

**Comment by ajkr** [comment]
I was also confused while reading the code why it is called twice in case of allow_2pc, can you make the first call only if !allow_2pc?

**Comment by ajkr** [comment]
Never mind, I understand now. The piece that was missing from my understanding was that min_log_num is transaction-aware, so it includes logs that have been flushed but contain uncommitted transactions. So you want to get min_log_num earlier such that flush can't happen between transaction prepare and commit, because then we might only copy the log containing the commit.

**Pull request closed without merging**